### ISSUE TITLE
完型填空

### ISSUE THREAD

**Issue by ouhongxu** [PaddlePaddle/ERNIE#616]
可以提供一个返回前n个适宜词的方法吗？万分感谢~！
下面方法仅返回一个答案，而且没有词检验（如果一个填空中包含两个[MASK]，第一个mask取字典中概率最高的那个字，第二个mask也取字典中概率最高的那个字，有时这两个mask拼接起来不是一个词）。

在最新开源的[ERNIE动态图模型](https://github.com/PaddlePaddle/ERNIE/tree/dygraph)当中，做完形填空将会更加简单方便.

继承`ErnieModelForPretraining`并改写`forward`方法，便可以实现简单的完形填空逻辑(仅限ernie1.0中文模型；enire2.0等英文模型因为没有发布mask_lm任务相关的weight文件所以还无法做到完形填空)；

请参考：
```python
import os
import numpy as np

import paddle.fluid as F
import paddle.fluid.layers as L
import paddle.fluid.dygraph as D
from ernie.modeling_ernie import ErnieModelForPretraining, ErnieModel
from ernie.tokenizing_ernie import ErnieTokenizer

class ErnieCloset(ErnieModelForPretraining):
    def __init__(self, *args, **kwargs):
        super(ErnieCloset, self).__init__(*args, **kwargs)
        del self.pooler_heads
    def forward(self, src_ids, *args, **kwargs):
        pooled, encoded = ErnieModel.forward(self, src_ids, *args, **kwargs)
        encoded_2d = L.gather_nd(encoded, L.where(src_ids == mask_id))
        encoded_2d = self.mlm(encoded_2d)
        encoded_2d = self.mlm_ln(encoded_2d)
        logits_2d = L.matmul(encoded_2d, self.word_emb.weight, transpose_y=True) + self.mlm_bias
        return logits_2d


if __name__ == "__main__":
    model_dir = 'ernie-1.0'

    D.guard().__enter__()
    tokenizer = ErnieTokenizer.from_pretrained(model_dir)
    rev_dict = {v: k for k, v in tokenizer.vocab.items()}
    rev_dict[tokenizer.pad_id] = '' # replace [PAD]
    rev_dict[tokenizer.sep_id] = '' # replace [PAD]
    rev_dict[tokenizer.unk_id] = '' # replace [PAD]

    @np.vectorize
    def rev_lookup(i):
        return rev_dict[i]

    ernie = ErnieCloset.from_pretrained(model_dir)
    ernie.eval()

    ids, _ = tokenizer.encode('戊戌变法，又称百日维新，是 [MASK] [MASK] [MASK] 、梁启超等维新派人士通过光绪帝进行 的一场资产阶级改良。')
    mask_id = tokenizer.mask_id
    print(ids)
    ids = np.expand_dims(ids, 0)
    ids = D.to_variable(ids)
    logits = ernie(ids).numpy()
    output_ids = np.argmax(logits, -1)
    seg_txt = rev_lookup(output_ids)
    print(seg_txt)
```

_Originally posted by @Meiyim in https://github.com/PaddlePaddle/ERNIE/issues/173#issuecomment-619491262_

**Comment by zhuzhf** [comment]
在这个case中`logits.shape == [3, vocab_size]`, 可以通过
```python
import itertools
for candidate in itertools.product(**logits.tolist()):
    check(candidate)
     ...
```
遍历所有组合并根据情况裁剪。

**Comment by ZTurboX** [comment]
该怎样加速模型推理速度呢
